File: backend/src/main/java/com/joying/chat/controller/PushNotificationController.java
package com.joying.chat.controller;

import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.joying.chat.dto.PushSubscriptionRequest;
import com.joying.chat.dto.PushUnsubscribeRequest;
import com.joying.chat.service.WebPushService;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;

/**
 * 브라우저 푸시 알림 구독을 다룬다.
 */
@RestController
@RequestMapping("/api/v1/push")
@RequiredArgsConstructor
public class PushNotificationController {

	private static final Logger log = LoggerFactory.getLogger(PushNotificationController.class);

	private final WebPushService webPushService;

	@Value("${web-push.public-key:}")
	private String vapidPublicKey;

	/**
	 * 구독을 만들 때 브라우저가 필요로 하는 공개키.
	 */
	@GetMapping("/vapid-public-key")
	public ResponseEntity<Map<String, String>> getVapidPublicKey() {
		return ResponseEntity.ok(Map.of("publicKey", vapidPublicKey));
	}

	@PostMapping("/subscribe")
	public ResponseEntity<Map<String, String>> subscribe(
		@Valid @RequestBody PushSubscriptionRequest request) {

		Long memberId = CurrentMember.id();
		log.info("푸시 구독 등록 요청: memberId={}, endpoint={}", memberId, request.getEndpoint());

		webPushService.subscribe(memberId, request);
		return ResponseEntity.ok(Map.of("message", "푸시 알림 구독이 등록되었습니다"));
	}

	@PostMapping("/unsubscribe")
	public ResponseEntity<Map<String, String>> unsubscribe(
		@Valid @RequestBody PushUnsubscribeRequest request) {

		webPushService.unsubscribe(request.getEndpoint());
		return ResponseEntity.ok(Map.of("message", "푸시 알림 구독이 해제되었습니다"));
	}
}

File: backend/src/main/kotlin/com/joying/chat/controller/PushNotificationController.kt
package com.joying.chat.controller

import com.joying.chat.dto.PushSubscriptionRequest
import com.joying.chat.dto.PushUnsubscribeRequest
import com.joying.chat.service.WebPushService
import jakarta.validation.Valid
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.ResponseEntity
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.bind.annotation.*

/**
 * 푸시 알림 컨트롤러
 */
@RestController
@RequestMapping("/api/v1/push")
class PushNotificationController(
    private val webPushService: WebPushService,
    @Value("\${web-push.public-key:}")
    private val vapidPublicKey: String
) {
    private val logger = LoggerFactory.getLogger(PushNotificationController::class.java)

    /**
     * VAPID 공개키 조회
     * 프론트엔드에서 Push 구독 시 필요
     */
    @GetMapping("/vapid-public-key")
    fun getVapidPublicKey(): ResponseEntity<Map<String, String>> {
        return ResponseEntity.ok(mapOf("publicKey" to vapidPublicKey))
    }

    /**
     * 푸시 구독 등록
     */
    @PostMapping("/subscribe")
    fun subscribe(
        @Valid @RequestBody request: PushSubscriptionRequest
    ): ResponseEntity<Map<String, String>> {
        val memberId = getCurrentMemberId()

        logger.info("푸시 구독 등록 요청: memberId={}, endpoint={}", memberId, request.endpoint)

        webPushService.subscribe(memberId, request)
        return ResponseEntity.ok(mapOf("message" to "푸시 알림 구독이 등록되었습니다"))
    }

    /**
     * 푸시 구독 해제
     */
    @PostMapping("/unsubscribe")
    fun unsubscribe(
        @Valid @RequestBody request: PushUnsubscribeRequest
    ): ResponseEntity<Map<String, String>> {
        webPushService.unsubscribe(request.endpoint)
        return ResponseEntity.ok(mapOf("message" to "푸시 알림 구독이 해제되었습니다"))
    }

    /**
     * SecurityContext에서 현재 인증된 사용자 ID 반환
     *
     * JwtAuthenticationFilter가 쿠키에서 JWT 토큰을 추출하고
     * SecurityContext에 인증 정보를 설정합니다.
     *
     * @return 사용자 ID
     * @throws IllegalStateException 인증되지 않은 경우
     */
    private fun getCurrentMemberId(): Long {
        val authentication = SecurityContextHolder.getContext().authentication
            ?: throw IllegalStateException("인증 정보가 없습니다")

        return authentication.name.toLong()
    }
}